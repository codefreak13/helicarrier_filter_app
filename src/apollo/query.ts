import {gql} from '@apollo/client';

export const GET_CHARACTERS = gql`
  query GetCharacters($page: Int, $filter: FilterCharacter) {
    characters(page: $page, filter: $filter) {
      results {
        id
        name
        status
        type
        gender
        origin {
          id
          name
        }
        created
      }
    }
  }
`;

interface Character {
  id: string;
  name: string;
  status: string;
  gender: string;
  origin: {name: string};
  created: string;
  type: string;
}

interface FilterCharacter {
  name?: string;
  status?: string;
  type?: string;
  gender?: string;
}

export interface CharacterVars {
  filter?: FilterCharacter;
  page?: number;
}

export interface CharacterData {
  characters: {
    results: Character[];
  };
}
