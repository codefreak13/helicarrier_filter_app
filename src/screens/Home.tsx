import React from 'react';
import {View, StyleSheet} from 'react-native';
import {Header, ListView, Loader} from '../components';
import {BoldText, Button, Input, PickerSelect} from '../ui';
import {COLORS, hp} from '../utils/Utils';
import {useBiodata} from '../hooks';
import {nameData, genderData, originData, statusData, typeData} from '../data';

const Home = () => {
  const {
    onChangeFormValue,
    clearSelection,
    renderedData,
    onChangeText,
    searchParam,
    onClearText,
    loading,
    gender,
    status,
    origin,
    error,
    name,
    type,
  } = useBiodata();

  return (
    <>
      {loading ? (
        <Loader />
      ) : error ? (
        <BoldText customstyle={styles.errorText}>Network Error</BoldText>
      ) : (
        <View style={styles.mainStyle}>
          <Header title="BIODATA" customMiddleIcon />
          <Input
            value={searchParam}
            placeholder="Search"
            setValue={onChangeText}
            onClear={onClearText}
          />
          <View style={styles.filterBar}>
            <PickerSelect
              items={nameData}
              value={name}
              placeholder="Name"
              onValueChange={onChangeFormValue('name')}
            />
            <PickerSelect
              items={statusData}
              value={status}
              placeholder="Status"
              onValueChange={onChangeFormValue('status')}
            />
            <PickerSelect
              items={genderData}
              value={gender}
              placeholder="Gender"
              onValueChange={onChangeFormValue('gender')}
            />
            <PickerSelect
              items={originData}
              value={origin}
              placeholder="Origin"
              onValueChange={onChangeFormValue('origin')}
            />
            <PickerSelect
              items={typeData}
              value={type}
              placeholder="Type"
              onValueChange={onChangeFormValue('type')}
            />
          </View>
          {(!!name || !!status || !!gender || !!origin || !!type) && (
            <Button
              title="Clear Selection(s)"
              onPress={clearSelection}
              customstyle={styles.buttonStyle}
            />
          )}
          <ListView data={renderedData} />
        </View>
      )}
    </>
  );
};

const styles = StyleSheet.create({
  mainStyle: {
    flex: 1,
    marginVertical: hp(20),
    alignItems: 'center',
    marginHorizontal: hp(15),
  },
  buttonStyle: {
    alignSelf: 'center',
    alignItems: 'center',
    borderWidth: hp(0.7),
    borderColor: COLORS.Black,
    backgroundColor: COLORS.White,
    borderRadius: hp(8),
    padding: hp(5),
  },
  errorText: {
    alignSelf: 'center',
    marginVertical: hp(50),
    color: COLORS.Danger,
  },
  filterBar: {
    flexDirection: 'row',
    marginVertical: hp(20),
  },
});

export default Home;
