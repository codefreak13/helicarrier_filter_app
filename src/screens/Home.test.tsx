import React from 'react';
import {render} from '@testing-library/react-native';
import {MockedProvider} from '@apollo/client/testing';
import Home from './Home';
import {GET_CHARACTERS} from '../apollo/query';

const mocks: any = [
  {
    request: {
      query: GET_CHARACTERS,
      variables: {page: 6, filter: {name: 'u'}},
    },
    results: {
      data: {
        characters: {
          result: [
            {
              id: '123',
              name: 'Mark',
              status: 'Alive',
              gender: 'Male',
              origin: {id: '11', name: 'Earth'},
              created: 'string',
              type: 'Pluto',
            },
          ],
        },
      },
    },
  },
];

describe('Home', () => {
  it('renders Home correctly', () => {
    const {toJSON, getByTestId} = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Home />
      </MockedProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
    expect(getByTestId('loader')).toBeTruthy();
  });
});
