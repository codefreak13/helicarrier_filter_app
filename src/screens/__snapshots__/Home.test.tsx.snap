// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders Home correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
  testID="loader"
>
  <ActivityIndicator
    color="#1D1F20"
    size="large"
  />
</View>
`;
